cpp/src/shuffler/chunk.cpp

```diff
@@ -87,6 +86,10 @@ std::unique_ptr<cudf::table> Chunk::unpack(rmm::cuda_stream_view stream) const {
     return unpack_and_concat(std::move(packed_vec), stream, br->device_mr());
 }
 
+bool Chunk::is_ready() const {
+    return !gpu_data || gpu_data->is_ready();
```

Comment:
There is a small nuance here. Chunk serves 2 purposes, 
1. Carrying messages
2. Flagging end-of-messages (by setting a non-zero value for `expected_num_chunks`). 

I think this should have the following logic, IINM. @madsbk WDYT?
```c++
   if (expected_num_chunks > 0){
      // always ready. Doesn't have any data buffers 
      return true;
   } else {
      // now, gpu_data buffer needs to be set at some point 
      if (gpu_data){
         return gpu_data->is_ready();
      } else {
         return false;
      } 
   }
```
This could reduce to, 
```c++
return (expected_num_chunks > 0) || (gpu_data && gpu_data->is_ready()); 
```

Reply:
But, I don't want to block the PR on this.

Reply:
```c++
return (expected_num_chunks > 0) || (gpu_data && gpu_data->is_ready());
```
Agree, this will make the intention more clear.

Reply:
I see this now, but it wasn't immediately obvious. I think it makes sense to remove `expected_num_chunks` from the public `Chunk` constructor that gets `gpu_data`, I've applied your suggestion and mine in https://github.com/rapidsai/rapidsmpf/pull/225/commits/de3db36003b3faaf38503b519e8661b6f1f5aa41, let me know if there's any reason not to make the change with constructors.

Reply:
Note that the private constructor remains there because it's used in `Chunk::from_metadata_message`.

Reply:
@nirandaperera pointed the condition from the commit mentioned above was wrong, it's now fixed in https://github.com/rapidsai/rapidsmpf/pull/225/commits/294d5b91bd4ec31b3a29de1c6a8013c7292ba33b.

Reply:
+1 on constructors. I think its much more clearer now. LGTM